Implement SIP-61 `@unroll` annotation

bishabosha opened the pull request:
~Still need to write documentation,~
Documentation is written also

I am doing this work on behalf of @lihaoyi 

The main implementation follows [com-lihaoyi/unroll](https://github.com/com-lihaoyi/unroll) but with some changes:
- `@unroll` annotation is `@experimental`
- run before pickling
- increase validation checks for incorrect usage of `@unroll`
- `Underscore` not `EmptyTree` in pattern match default case
- ensure correct spans in TASTy
- ensure symbols are correctly substituted in types

There is one main library addition: `scala.annotation.unroll`, i.e. the `@unroll` annotation that appears on parameters,

commits are a bit of a mess - a lot of churn

Edit: removed changes:
- ~reuse the symbol when generating the "forwarder" of an abstract method.~
- ~infer `override` when overriding trait methods that "appear" abstract in source code, but were actually implemented by the unroll annotation~
- ~do not add the `Invisible` flag to abstract unrolled methods - this means they can actually still be visible in separately compiled compilation units (because unrolling runs before pickler now)~
- ~Internal annotation `scala.annotation.internal.AbstractUnroll` - marker used to indicate that an unrolled abstract method was abstract in source code (unrolling phase actually provides a default implementation) - this enables downstream implementers to avoid the override flag. (Maybe we shouldn't allow this convenience?)~
- ~Internal annotation `scala.annotation.internal.UnrollForwarder` - marker on any generated forwarder - which is used for special casing TASTy pickling of method calls to the forwarders~
- ~because forwarders get the `Invisible` flag, this makes them impossible to select from TASTy normally (via `SELECTin`), (invisible def are filtered out "before typer") so I intercept the `Select` trees to be `TERMREFdirect`, and then restore them to Select after TASTY. perhaps someone has a better idea, or we could change the resolution rules for `Invisible`? or invent a new TASTy node? (I also tried generating a `Ident` tree rather than a `Select`, but this had a type error)~


fixes #21728

lihaoyi commented:
@bishabosha some high level comments:

1. We ended up removing the `abstract def` support from the SIP (https://github.com/scala/improvement-proposals/pull/78) since we weren't confident in the semantics, so let's disable it in the implementation as well as raise an error

2. It's not obvious to me from the tests, but is it possible to raise errors for the following cases:
    * `@unroll` on an invalid statement (e.g. on a `val`, or on a type `foo: @unroll`
    * `@unroll` on a non-final class/trait method: `@unroll` and its delegation model generally only works with final methods, and cannot support overrides with things getting wonky

3. The original compiler plugin didn't have support for `trait` parameter lists. Would that be easy to add? It's probably not critical, but would be nice to have for consistency

bishabosha commented:
Right, I should have noted that the abstract method support should have been dropped - now it is, I will push further commits with more invalidation checks, and see if trait constructor unroll can work

bishabosha commented:
@lihaoyi there isn't a way to support traitParameters that isn't a straightforward rewrite that a user could do manually, so I think this is unexplored territory- e.g. providing default implementations for param accessors of traits

lihaoyi commented:
> @lihaoyi there isn't a way to support trait constructor parameters that isn't a rewrite that a user could do manually, so I think this is unexplored territory - e.g. providing default implementations in bytecode for param accessors of traits
> 
> All the other concerns were addressed in above commits

Sounds good, let's skip trait params for now.

Anyone from the Scala 3 side able to review the code itself and the integration into the scala 3 codebase?

bishabosha commented:
one implication I guess is that someone might not plan to use `@unroll` when they first made the API, so not make the method final, and then are potentially restricted from introducing `@unroll` in the future. I guess we should include in our documentation that you should make (all?) methods final

lihaoyi commented:
I think forcing `final` is probably necessary, either explicitly by asking the user to put the keyword, or implicitly by having the compiler add it. 

At least as currently designed, if code can override an `@unroll`ed method, it can result in different callsites running different logic depending on what version of the `@unroll`ed API they were compiled against, which violates all sorts of expectations.

So for `case class`es and `object` methods and constructors, things are already `final` so the requirement doesn't make a difference. 

For `class` and `trait` methods, if they're not final but nobody is overriding them, then having the upstream API add `final` changes nothing. If they are not final but someone *is* overriding them, then adding `final` upstream is the difference between a loud JVM linkage error and a silent misbehavior, and I think the loud JVM linkage error is the preferable failure mode

lihaoyi commented:
CC @lrytz since you're the manager, not sure if you should be reviewing this?

bishabosha commented:
@Gedochao I though experimental stuff can go in a patch?

Gedochao commented:
> @Gedochao I thought experimental stuff can go in a patch? - I guess the tasty peculiarity - but that will have experimental flag in the tasty if used

Ah, if it's behind a flag then it's fine, my bad.

bishabosha commented:
@Gedochao I added it back - there would be no block at the tasty level if this were in a hypothetical 3.5.2 release, so it should be in 3.6.0 minimum

bishabosha commented:
Just added in the documentation

bishabosha commented:
In commit https://github.com/scala/scala3/pull/21693/commits/b0ece26b38066e02efaf262462903aacd7ab4b9d I changed the way forwarders are generated - i.e. they all call the original unrolled method. This means that the TASTy hack is no longer needed.

bishabosha commented:
I was talking to @sjrd today, and there is still a problem with `Invisible` - we have to consider the case of an inline method that calls a method, and then in V2 they add an unrolled parameter. That TASTy will still try to resolve the old method by name+signature (i.e. try to resolve the invisible forwarder) and fail.

lihaoyi commented:
How about if we leave the synthetic overloads visible. They're meant to be equivalent semantically - it should not matter which one ends up getting called. That's the current manual status quo and it works well enough.

sjrd commented:
That would definitely have user-visible consequences. For starters, type inference gets worse as soon as a method has overloads: you don't get expected types, so you don't get lambda param type inference, and implicit conversions may be inserted later or sooner than otherwise.

If we need to go that far, it's a language design change compared to the SIP, so we need to go back there. But I don't think we need to do that; if `Invisible` does not have the semantics we want, we can introduce a new TASTy flag with the semantics we want.

lihaoyi commented:
A new flag sounds reasonable if it's possible on the implementation side

bishabosha commented:
I added a test that fails to resolve a forwarder when inlining a transparent inline method, then I fixed the test by introducing a new SourceInvisible flag that is identical to Invisible, except in the case of resolving SELECTin from TASTy

bishabosha commented:
I think we should probably think carefully about the intention and semantics of `SourceInvisible` - it might have to play into macro checking as well - e.g. it allows definitions to go from visible to invisible

bishabosha commented:
So the conclusion from the core meeting was to go ahead and adapt `Invisible` to match the semantics of `SourceInvisible`

tgodzik commented:
> So the conclusion from offline discussion was to go ahead and adapt `Invisible` to match the semantics of `SourceInvisible`

That's right, if nothing breaks we should be fine. And I think only `@main` uses it anyway

bishabosha commented:
> That's right, if nothing breaks we should be fine. And I think only `@main` uses it anyway

it should also be bean getters and I believe any macro annotation generated code

bishabosha commented:
@sjrd I added the new "visible from TASTy" semantic for `Invisible` (removing `SourceInvisible`) - is that a "bug fix" or a needs-minor-release?

bishabosha commented:
@WojciechMazur its seems according to @tgodzik this needs to be tested with Open Community Build to check for breakages caused by the new behaviour of the invisible flag

tgodzik commented:
> @sjrd @Gedochao @tgodzik I added the new "visible from TASTy" semantic for `Invisible` (removing `SourceInvisible`) - is that a "bug fix" or a `needs-minor-release`?

Not sure, @Gedochao is back on Wednesday we will discuss it then.

WojciechMazur commented:
> does this need to be tested with Open Community Build to check for breakages caused by the new behaviour of the invisible flag? 

I've started the build to test it out

WojciechMazur commented:
There seems to be no new regressions when comparing with latest nightly 3.6.1-RC1-bin-20241011-5e7ab15-NIGHTLY

lihaoyi commented:
Just to follow up here since the community build was clean, is there anything else we need to do on the compiler side?

Gedochao commented:
> is there anything else we need to do on the compiler side?

@lihaoyi We'll discuss this on the next Scala Core meeting and get back to you by the end of the week.

Gedochao commented:
It's been decided we should try to get this ready for 3.7.x.
So the next step would be to wait for it to get reviewed.
cc @lihaoyi @bishabosha @sjrd @lrytz

lihaoyi commented:
Thanks @Gedochao! Will wait for @sjrd  and @lrytz to take a look

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+
+  import tpd.*
+
+  private var _unrolledDefs: util.EqHashMap[Symbol, ComputedIndicies] | Null = null
+  private def initializeUnrolledDefs(): util.EqHashMap[Symbol, ComputedIndicies] =
```
Here again, I don't see why we're trying so hard to avoid allocating a unique instance of `EqHashMap`.

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+  def newTransformer(using Context): Transformer =
+    UnrollingTransformer(ctx.compilationUnit.unrolledClasses.nn)
+
+  type ComputedIndicies = List[(Int, List[Int])]
+  type ComputeIndicies = Context ?=> Symbol => ComputedIndicies
```
```suggestion
  type ComputedIndices = List[(Int, List[Int])]
  type ComputeIndices = Context ?=> Symbol => ComputedIndices
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+    def computeIndices(annotated: Symbol)(using Context): ComputedIndicies =
+      unrolledDefs.getOrElseUpdate(annotated, {
+        if annotated.name.is(DefaultGetterName) then
+          Nil // happens in curried methods where more than one parameter list has @unroll
```
The doc says

> `@unroll` may be applied to more than one parameter per method, but all occurrences must appear in the same parameter clause.

So this shouldn't happen. Or does it happen for invalid programs that survive through this phase?

bishabosha replied:
The comment is incorrect, and should be updated, the problem is from this:
```scala
final def foo(@unroll y: String = "")(x: Int = 23) = x
```
the default getter for `x` will be 
```scala
def foo$default$2(@unroll y: String): Int @uncheckedVariance = 23
```
and we should ignore generating forwarders for this

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+  type ComputedIndicies = List[(Int, List[Int])]
+  type ComputeIndicies = Context ?=> Symbol => ComputedIndicies
+
+  private class UnrollingTransformer(classes: Set[Symbol]) extends Transformer {
```
Consider giving a better name to the `classes` parameter. Here it's decent but at use site it's confusing what it's supposed to be.

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+        super.transform(tree)
+  }
+
+  def copyParamSym(sym: Symbol, parent: Symbol)(using Context): (Symbol, Symbol) =
```
Make all these methods `private`?

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+
+  def isTypeClause(p: ParamClause) = p.headOption.exists(_.isInstanceOf[TypeDef])
+
+  def generateSingleForwarder(defdef: DefDef,
```
Consider adding a comment to this method to show what it's supposed to generate. Notably what its 8 (!) parameters mean and how they influence the result.

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+        defdef.symbol.flags &~ HasDefaultParams |
+        Invisible | Synthetic,
+        NoType, // fill in later
+        coord = nextSymbol.span.shift(1) // shift by 1 to avoid "secondary constructor must call preceding" error
```
:confounded: Can we do this better? Like not reporting "secondary constructor must call preceding" for `Synthetic` callers?

bishabosha replied:
I was trying to limit the footprint across the compiler until review - this extra check works, hope no tests break

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+      val (oldParams, newParams) = newParamSymMappings.flatten.unzip
+
+      val newParamSymLists0 =
+        newParamSymMappings.map: pairss =>
+          pairss.map: (oldSym, newSym) =>
```
`pairss` doesn't appear to be a list of lists, is it? If not, call it `pairs`:
```suggestion
        newParamSymMappings.map: pairs =>
          pairs.map: (oldSym, newSym) =>
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+          else ps.map(ref)
+        }
+
+      val forwarderCall0 = forwarderCallArgs.foldLeft[Tree](forwarderInner){
+        case (lhs: Tree, newParams) =>
+          if (newParams.headOption.exists(_.isInstanceOf[TypeTree])) TypeApply(lhs, newParams)
+          else Apply(lhs, newParams)
+      }
```
The same logic can already be fond on line 197. Factor out? In fact, isn't there already an overload of `appliedTo` that handles this?

bishabosha replied:
good catch - was from copied code

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+                )
+            )
+          )
+        } ++ Seq(
```
```suggestion
        } :+ (
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+      import dotty.tools.dotc.core.NameOps.isConstructorName
+
+      val isCaseCopy =
+        defdef.name.toString == "copy" && defdef.symbol.owner.is(CaseClass)
```
Use `nme.copy`. Same for `"apply"` and `"fromProduct"` below.

bishabosha replied:
oops, left over from copying the code

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+    case _ => None
+  }
+
+  def unrollTemplate(tmpl: tpd.Template, compute: ComputeIndicies)(using Context): tpd.Tree = {
```
```suggestion
  def unrollTemplate(tmpl: tpd.Template, computeIndices: ComputeIndices)(using Context): tpd.Tree = {
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+    ).setDefTree
+  }
+
+  def generateSyntheticDefs(tree: Tree, compute: ComputeIndicies)(using Context): Option[(Symbol, Option[Symbol], Seq[DefDef])] = tree match {
```
```suggestion
  def generateSyntheticDefs(tree: Tree, computeIndices: ComputeIndices)(using Context): Option[(Symbol, Option[Symbol], Seq[DefDef])] = tree match {
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+                    paramIndex,
+                    paramCount,
+                    nextParamIndex,
+                    nextSymbol,
```
Is there a good reason that we don't make all the forwarders directly call `defdef.symbol`? That should be simpler and potentially more efficient, wouldn't it?

bishabosha replied:
I'd agree, this was a leftover from the mode that supported abstract methods

bishabosha replied:
the argument made was byte code reduction, its also in the SIP text if that matters, or can that be regarded as implementation detail?

lihaoyi replied:
Either way works, one has a larger bytecode size and one has a larger runtime overhead. I chose the larger runtime overhead option, but semantically both should be identical

lihaoyi replied:
yeah also it was relied upon by the abstract method stuff that is now removed

bishabosha replied:
actually, the implementation does call the def directly, nextSymbol is only being used to compute that shifted span offset - I'll try to avoid needing that

sjrd commented on compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala:
```diff
@@ -0,0 +1,358 @@
+    val bodySubs = generatedBody.flatMap((_, maybeSub, _) => maybeSub).toSet
+    val otherDecls = tmpl.body.filterNot(d => d.symbol.exists && bodySubs(d.symbol))
+
+    /** inlined from compiler/src/dotty/tools/dotc/typer/Checking.scala */
```
Hum, maybe call the original? Or factor out what should be factored out?

sjrd commented:
Regarding tests:

* Do all the sbt tests need to be sbt tests? It seems to me that most of them would fit into Vulpix' support for separate compilation, wouldn't they? That would be faster + it would also test under Scala.js.
* We need tests with `AnyVal`s. Overloads in `AnyVal`s get weird suffixes that are position-dependent. This is likely to cause issues with this feature if we're not extra careful.
* It would be good to have one test that uses `@unroll` on a clause-interleaved method.

bishabosha commented:
> Do all the sbt tests need to be sbt tests? It seems to me that most of them would fit into Vulpix' support for separate compilation, wouldn't they?

probably, it required the least work to migrate from the unroll GitHub repo the way it was done

bishabosha commented on compiler/src/dotty/tools/dotc/CompilationUnit.scala:
```diff
@@ -59,6 +59,9 @@ class CompilationUnit protected (val source: SourceFile, val info: CompilationUn
 
   var hasMacroAnnotations: Boolean = false
 
+  def hasUnrollDefs: Boolean = unrolledClasses != null
+  var unrolledClasses: Set[Symbol] | Null = null
```
good point

bishabosha commented on compiler/src/dotty/tools/dotc/core/Denotations.scala:
```diff
@@ -1074,7 +1074,7 @@ object Denotations {
       if (denots.exists && denots.matches(this)) NoDenotation else this
     def filterWithFlags(required: FlagSet, excluded: FlagSet)(using Context): SingleDenotation =
-      val realExcluded = if ctx.isAfterTyper then excluded else excluded | Invisible
+      val realExcluded = if ctx.isAfterTyper then excluded else excluded | (if ctx.mode.is(Mode.ResolveFromTASTy) then EmptyFlags else Invisible)
```
yep

bishabosha commented on compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:
yeah, this would be leftover from when I previously changed the tasty encoding

bishabosha commented on compiler/src/dotty/tools/dotc/inlines/Inlines.scala:
```diff
@@ -158,8 +159,18 @@ object Inlines:
+                EmptyTree
+              else
+                // Tested in sbt-test/tasty-compat/add-param-unroll2/a_v3_2/C.scala
+                ctx.compilationUnit.suspend("suspending in case of possible generated methods")
```
This analysis of the tree picker case is correct - the root issue is anytime some tasty is inlined (transparent inline) that references an "old" signature of method `foo`, where foo is defined in the same compilation run, so therefore forwarders will not be generated yet (before `foo` is even analyzed to see if it has `unroll`)

bishabosha replied:
this is the error if you do nothing to detect the scenario:
```scala
[error] a_v2/B.scala:7:35: wrong number of arguments at parser for (s: String, x: Int, b: Boolean): String: (a.A.foo : (s: String, x: Int, b: Boolean): String), expected: 3, found: 2
[error]   transparent inline def caller = A.foo("abc", 2)
```

bishabosha replied:
perhaps we can produce a generic error whenever TASTy SELECTin does not resolve a matching method signature - like `Foo.tasty is broken... could not resolve matching `foo` from class path`

bishabosha replied:
Removing this code - in another PR we can introduce a generic solution to incorrectly resolving SELECTin nodes

bishabosha commented on compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:
```diff
@@ -170,11 +170,13 @@ class TreeUnpickler(reader: TastyReader,
+    def forkAt(start: Addr, inInlineBody: Boolean = false): TreeReader =
+      new TreeReader(subReader(start, endAddr), inInlineBody)
+
+    def fork: TreeReader = forkAt(currentAddr, inInlineBody)
```
removed - we can deal with incorrectly resolving SELECTin in another PR

bishabosha commented:
thanks to automation moving 129 files is easy https://gist.github.com/bishabosha/1baf30b4b38cf9e9795e8664c64bc7c5

bishabosha commented:
CI passing again

bishabosha commented:
@sjrd rebased

lihaoyi reviewed (approved):
I can't really review the implementation, but skimming the test suite it seems to do what it says on the box and looks good to me

bishabosha commented:
ci error due to a diff file changing

bishabosha commented:
Green ci again

lihaoyi commented:
Bumping reviewers @sjrd @lrytz, since it's been a few weeks

lrytz reviewed (approved):
Looks great! I didn't look at `transform/UnrollDefinitions.scala` in detail.

Is there anything that should be done in MiMa, or in tasty-mima?

bishabosha commented:
I'll work on the feedback tonight

bishabosha commented on compiler/src/dotty/tools/dotc/reporting/messages.scala:
```diff
@@ -3350,7 +3350,6 @@ final class DeprecatedAssignmentSyntax(key: Name, value: untpd.Tree)(using Conte
       |not as an assignment.
       |
       |To assign a value, use curly braces: `{${key} = ${value}}`."""
-      + Message.rewriteNotice("This", version = SourceVersion.`3.6-migration`)
```
I'll add this back, its probably a mistake from fixing a git conflict

bishabosha commented on library/src/scala/annotation/unroll.scala:
```diff
@@ -0,0 +1,4 @@
+package scala.annotation
+
+@experimental("under review as part of SIP-61")
+final class unroll extends scala.annotation.StaticAnnotation
```
good catch, thanks!

bishabosha commented on tests/run/unroll-traitMethod-integration/UnrollTestPlatformSpecific_3.scala:
```diff
@@ -0,0 +1,30 @@
+        "hello2false3"
+    )
+
+    cls.getMethods.filter(_.getName.contains("foo")).foreach(println)
```
there was some flakiness, but now its fixed with sorting, so I added them in

bishabosha commented on tests/run/unroll-traitMethod/Test_2.scala:
```diff
@@ -0,0 +1,6 @@
+@main def Test =
+
+  // ensure that impl.foo isn't `Invisible`, so can be resolved from TASTy
+  assert(UnrolledImpl.impl.foo("foo") == "foo1true")
```
removing this test - its duplicated anyway in the unroll-traitMethod-integration which inspects via reflection.

bishabosha commented on tests/run/unroll-traitMethod-integration/Test_4.scala:
```diff
@@ -0,0 +1,8 @@
+//> using options -experimental
+// scalajs: --skip
```
im working on adding `//> using target.platform (jvm|scala-js)` to the vulpix runner, which should allow to do this (and have separate check files)

bishabosha replied:
added in https://github.com/scala/scala3/pull/21693/commits/b894262d27c4bf70dbe93fb7562196f96037072b

bishabosha commented:
I've added in https://github.com/scala/scala3/pull/21693/commits/b894262d27c4bf70dbe93fb7562196f96037072b support for `//> using target.platform jvm|scala-js` in vulpix tests - which filter out files not matching the "test platform" - which also allows to have a separate check file for each platform.

If this is acceptable, should it be lifted to a separate PR?

bishabosha commented on compiler/src/dotty/tools/dotc/typer/CrossVersionChecks.scala:
```diff
@@ -81,18 +81,37 @@ class CrossVersionChecks extends MiniPhase:
+    if annotSym == defn.UnrollAnnot then
+      unrollError(pos)
+
+  private def checkUnrollMemberDef(memberDef: MemberDef)(using Context): Unit =
```
not really - first there is no method for inspecting annotations in a miniphase, and second no common (`prepareForSymbol`) so you really have to handle each kind of tree separately

bishabosha commented:
@lrytz All comments addressed - CI green
@Gedochao who else should look at this?

Gedochao commented:
@bishabosha we'll discuss this on this week's Scala Core

Gedochao commented:
This is planned to be included in 3.7.0.
Waiting for one last round of review (cc @sjrd)

lrytz commented:
> Is there anything that should be done in MiMa, or in tasty-mima?

@bishabosha ping on that question

bishabosha commented:
> Is there anything that should be done in MiMa, or in tasty-mima?

for MiMa, the forwarders should look like ordinary scala methods, and will be present in TASTy/bytecode as usual.

There is the new semantic of the `Invisible` flag which is that it should be visible when resolved from an inline method.

sjrd commented:
I don't think we need anything in tasty-mima. tasty-query ignores the `Invisible` flag, which corresponds to the new semantics, as far as it is concerned.

sjrd commented:
@bishabosha Is the git history relevant? Or should it all be squashed to help future `git blame`s?

bishabosha commented:
> Is the git history relevant? Or should it all be squashed to help future git blames?

there is perhaps 3-commit split - the initial commit which is copy-paste the com-lihaoyi/unroll code, commit 2 which is "the rest of the work" and perhaps a third commit which adds the vulpix directive parsing stuff

sjrd commented:
I'll let you decompose as you see fit. Then we can merge.

bishabosha commented:
@sjrd I squashed the commits

unkarjedy commented:
Hello!
Is it fine if I leave some feedback directly here?

I tried 3.7.0-RC4

The error ranges seem inconsistent & strange to me
### Example 1
"Cannot unroll method foo in class Example at line 5 because parameter z needs a default value"
The error is specific to a single parameter.
The whole parameter range is highlighted.
OK.
```scala
import scala.annotation.unroll

class Example {
    final def foo(x: Int, y: Int, @unroll z: Int): Int = x + y + z
}
```
<img width="907" alt="image" src="https://github.com/user-attachments/assets/4fc94381-fb22-4293-8591-66b6f30df807" />

### Example 2
"Cannot unroll parameters of method foo: it is not final"
The error is specific to the whole function and the @unroll annotation usage fact.
However, the range here is the parameter name identifier, which looks strange.
Nothing in the parameter `z` is special here, so I suppose the better range to highlight here would be just `@unroll` annotation

```scala
import scala.annotation.unroll

class Example {
    def foo(x: Int, y: Int, @unroll z: Int): Int = x + y + z
}
```
<img width="674" alt="image" src="https://github.com/user-attachments/assets/a07df74e-863c-46a7-8d5c-bc984054e1ea" />

### Example 3
"@unroll is only allowed on a method parameter"
It's shown on the name identifier `foo`.
To me it seems the `@unroll` annotation itself would be a better fit here as well - it's the annotation to blame, not the definition name identifier.
```scala
import scala.annotation.unroll

class Example {
  @unroll def foo(x: Int, y: Int, z: Int): Int = x + y + z
}
```
<img width="521" alt="image" src="https://github.com/user-attachments/assets/e664c591-1cba-4b57-9f7d-d58fad3d52ec" />
